casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala

```diff
@@ -58,32 +63,111 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
-        ConstructDeploy.sourceDeploy(echoContract(2), time + 2, accounting.MAX_VALUE)
-      )
-      for {
-        block0 <- nodes(0).addBlock(deploys(0))
-        block1 <- nodes(1).addBlock(deploys(1))
-        _      <- nodes(0).receive()
-        _      <- nodes(1).receive()
-        _      <- nodes(0).receive()
-        _      <- nodes(1).receive()
+    checkMerge("Nil", echoContract(1), echoContract(2))
+  }
 
-        //multiparent block joining block0 and block1 since they do not conflict
-        multiparentBlock <- nodes(0).addBlock(deploys(1))
-        _                <- nodes(1).receive()
+  // TODO: Peek rows/column
+  // Note this skips pairs that lead to infinite loops
+  it should "handle multi-parent blocks correctly when they operate on volatile produce/consume pairs" in effectTest {
+    // 4! 4!
+    checkConflicts("@0!(0)", "for (_ <- @0) { 0 }", "for (_ <- @0) { 0 }") >>
+      // 4X 4X (TODO: Eventually should merge)
+      checkConflicts("0", "for (_ <- @0) { 0 }", "for (_ <- @0) { 0 }") >>
+      //
+      // !X !X (TODO: Eventually should merge)
+      checkConflicts("0", "@0!(0)", "@0!(0)") >>
+      // !X !4
+      checkMerge("for (_ <- @0) { 0 }", "@1!(0)", "@0!(0)") >>
+      // !X !C
+      checkMerge("contract @0(id) = { 0 }", "@1!(0)", "@0!(0)") >>
+      // !X 4X
+      checkConflicts("0", "@0!(0)", "for (_ <- @0) { 0 }") >>
+      // !X 4!
+      checkMerge("@0!(0)", "@1!(0)", "for (_ <- @0) { 0 }") >>
+      // !X 4!!
+      checkMerge("@0!!(0)", "@1!(0)", "for (_ <- @0) { 0 }") >>
+      // !X !!X (TODO: Eventually should merge)
+      checkConflicts("0", "@0!(0)", "@0!!(0)") >>
+      // !X !!4
+      checkMerge("for (_ <- @0) { 0 }", "@1!(0)", "@0!!(0)") >>
+      // !X CX
+      checkConflicts("0", "@0!(0)", "contract @0(id) = { 0 }") >>
+      // !X C!
+      checkMerge("@0!(0)", "@1!(0)", "contract @0(id) = { 0 }") >>
+      // (!4) (!4)
+      checkMerge("0", "for (_ <- @0) { 0 } | @0!(1)", "for (_ <- @0) { 0 } | @0!(2)") >>
+      // !4 !4
+      checkConflicts("for (_ <- @0) { 0 }", "@0!(1)", "@0!(2)") >>
+      // !4 !4
+      checkMerge("for (@1 <- @0) { 0 }", "@0!(1)", "for (@2 <- @0) { 0 } | @0!(2)") >>
+      // !4 !C
+      checkMerge("for (@1 <- @0) { 0 } | contract @1(id) = { 0 }", "@0!(1)", "@1!(1)") >>
+      // !4 4X
+      checkMerge("for (_ <- @0) { 0 }", "@0!(0)", "for (_ <- @1) { 0 }") >>
+      // Skip !4 4! as handled above
+      // !4 4!!
+      checkMerge("for (_ <- @0) { 0 } | @1!!(0)", "@0!(0)", "for (_ <- @1) { 0 }") >>
+      // !4 !!X
+      checkMerge("for (_ <- @0) { 0 }", "@0!(0)", "@1!!(0)") >>
+      // !4 !!4
+      checkConflicts("for (_ <- @0) { 0 } | for (_ <- @0) { 0 }", "@0!(0)", "@0!!(0)") >>
+      // !4 CX (TODO: Check if this could conflict)
+      checkMerge("for (_ <- @0) { 0 }", "@0!(1)", "contract @1(id) = { 0 }") >>
+      // !4 C!
+      checkMerge("for (_ <- @0) { 0 } | for (_ <- @1) { 0 }", "@0!(0)", "contract @1(id) = { 0 }")
```

Comment:
I didn't mean for you to go through the whole sheet :) It's good to have more tests though.

At the same time, we should for now probably focus the combinations:
`{!X, !4, !C , 4X, 4!}^2`, maybe extending it to `{!X, !4, !C , 4X, 4!, CX, C!}^2`

Reply:
Notice there's no `!!` in the combinations above. Let's keep the tests for `!!` you've added, but not add more unless you see a need.